pulsar-broker/src/main/java/org/apache/pulsar/broker/service/DrainingHashesTracker.java

```diff
@@ -179,31 +221,43 @@ public DrainingHashesTracker(String dispatcherName, UnblockingHandler unblocking
-            if (log.isDebugEnabled()) {
-                log.debug("[{}] Adding and incrementing draining hash {} for consumer id:{} name:{}", dispatcherName,
-                        stickyHash, consumer.consumerId(), consumer.consumerName());
-            }
-            entry = new DrainingHashEntry(consumer);
-            drainingHashes.put(stickyHash, entry);
-            // update the consumer specific stats
-            consumerDrainingHashesStatsMap.computeIfAbsent(new ConsumerIdentityWrapper(consumer),
-                    k -> new ConsumerDrainingHashesStats()).addHash(stickyHash);
-        } else if (entry.getConsumer() != consumer) {
-            throw new IllegalStateException(
-                    "Consumer " + entry.getConsumer() + " is already draining hash " + stickyHash
-                            + " in dispatcher " + dispatcherName + ". Same hash being used for consumer " + consumer
-                            + ".");
-        } else {
-            if (log.isDebugEnabled()) {
-                log.debug("[{}] Draining hash {} incrementing {} consumer id:{} name:{}", dispatcherName, stickyHash,
-                        entry.refCount + 1, consumer.consumerId(), consumer.consumerName());
+
+        DrainingHashEntry entry;
+        ConsumerDrainingHashesStats addedStats = null;
+        lock.writeLock().lock();
+        try {
+            entry = drainingHashes.get(stickyHash);
+            if (entry == null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Adding and incrementing draining hash {} for consumer id:{} name:{}",
+                            dispatcherName, stickyHash, consumer.consumerId(), consumer.consumerName());
+                }
+                entry = new DrainingHashEntry(consumer);
+                drainingHashes.put(stickyHash, entry);
+                // add the consumer specific stats
+                addedStats = consumerDrainingHashesStatsMap.computeIfAbsent(new ConsumerIdentityWrapper(consumer),
+                        k -> new ConsumerDrainingHashesStats());
+            } else if (entry.getConsumer() != consumer) {
+                throw new IllegalStateException(
+                        "Consumer " + entry.getConsumer() + " is already draining hash " + stickyHash
+                                + " in dispatcher " + dispatcherName + ". Same hash being used for consumer " + consumer
+                                + ".");
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Draining hash {} incrementing {} consumer id:{} name:{}", dispatcherName,
+                            stickyHash, entry.getRefCount() + 1, consumer.consumerId(), consumer.consumerName());
+                }
             }
+        } finally {
+            lock.writeLock().unlock();
+        }
+        if (addedStats != null) {
+            // add hash to added stats
+            addedStats.addHash(stickyHash);
```

Comment:
Is this required? It seems like this is no-op as addedStats is a local var in this func.

Reply:
Yes, this is necessary. The stats instance has a separate lock. By making the call outside of the lock of the draining hashes instance, it reduces the chances for deadlocks. The deadlock that this PR is fixing was caused by such an access pattern in the reduceRefCount method.